Default Output Directory

natecraddock opened the pull request:
This adds a new preference for setting the default output directory for various operations within zygrader.
* Grade Puller
* Unmatched students
* Submission Search
* Student code i/o

For this to work, we set the zygrader current working directory to the path on startup, and whenever it is modified in the preferences.

Closes #33

natecraddock committed:
Preferences: Add Default Output Directory preference

Add a preference for specifying the starting directory for
file output. When text boxes for path input are shown, this directory
can be used for the starting path. (string match output, csv files).

We can also use it for setting the working directory for code
execution.

Part of #33

natecraddock committed:
Use default output directory for admin menu

Output files to the default output directory in the admin menu options
* Grade Puller
* Find unmatched students
* Submission Search

natecraddock committed:
Use default directory for zygrader working directory

Set zygrader's working directory to the user-set default
output directory. Input and output files for student code are read
and written from this directory now.

natecraddock committed:
Merge branch 'master' into default-output-dir

teiknsmith commented on zygrader/config/versioning.py:
```diff
@@ -97,6 +97,9 @@ def versioning_update_preferences():
     if compare_versions(version, user_version):
         config["spooky_mode"] = False
 
+    # Code here will always be run until the next version bump
+    config["output_dir"] = "~"
+
```
Shouldn't this still be in a conditional? Even if this function is only called when there are updates, it will override this preference for any new update.

teiknsmith replied:
Also, this could make use of `preferences.DEFAULT_PREFERENCES`. Perhaps there should be an unconditional check for non-existent preferences that sets them to the defaults.

natecraddock replied:
For development purposes it is nice to have an "always run this" section in versioning. Without it, you can run into versioning issues:
1. Feature A is added and bumps the version (say to 5.0.1) because a new default needs to be set
2. Time passes (zygrader is run)
3. Feature B is added and also needs versioning, but because the config version number is 5.0.1 the versioning code never runs.

So it's helpful to have a section like this that is always run until that version is officially released. With that release we move it to a conditional.

---

However, in this case we should probably make an exception for adding new preferences with a simple loop in the default preferences.

teiknsmith replied:
Ah, makes sense.

teiknsmith commented on zygrader/config/preferences.py:
```diff
@@ -35,6 +35,7 @@
     "class_code": "No Override",
     "editor": "Pluma",
     "data_dir": "",
+    "output_dir": "~",
```
I think a default of `~/` would be better from my testing of where this is used.

teiknsmith replied:
```suggestion
    "output_dir": "~/",
```

natecraddock replied:
Good idea, though I'm using os.path.join which should add the `/` by default

teiknsmith replied:
It's mostly just when presenting the default dir without an attached file, it was convenient to already have the `/` in the name and save a keystroke. (e.g. in Admin > Submissions Search > after lab select and search string, in Output File)

natecraddock committed:
Versioning: Add function to automatically add new preferences

Instead of manually adding new user preferences, automatically update the
user's config.json file with any new preference data added in the new
version.
